plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,321 @@
+    languageRegister = mkOption {
+      type = with types; attrsOf (coercedTo str toList (listOf str));
+      default = { };
+      example = {
+        cpp = "onelab";
+        python = [
+          "foo"
+          "bar"
+        ];
       };
+      description = ''
+        This is a wrapping of the `vim.treesitter.language.register` function.
+
+        Register specific parsers to one or several filetypes.
+
+        The keys are the parser names and the values are either one or several filetypes.
+      '';
+    };
 
-      nixvimInjections = mkEnableOption "nixvim specific injections, like lua highlighting in extraConfigLua";
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
+      description = "Whether to install grammars defined in `grammarPackages`.";
+    };
+
+    nixvimInjections = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
+      description = "Whether to enable Nixvim injections, e.g. highlighting `extraConfigLua` as lua.";
+    };
+
+    nodejsPackage = helpers.mkPackageOption {
+      name = "nodejs";
+      default = pkgs.nodejs;
+      defaultText = literalExpression "pkgs.nodejs";
+      example = literalExpression "pkgs.nodejs_22";
+      description = ''
+        Which package (if any) to be added as the nodejs package.
+
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of NodeJS, set this option to `null`.
+      '';
+    };
+
+    treesitterPackage = helpers.mkPackageOption {
+      name = "tree-sitter";
+      default = pkgs.tree-sitter;
+      defaultText = literalExpression "pkgs.tree-sitter";
+      description = ''
+        Which package (if any) to be added as the tree-sitter binary.
+
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of tree-sitter, set this option to `null`.
+      '';
     };
   };
 
-  config =
-    let
-      tsOptions = {
-        highlight = {
-          inherit (cfg) enable;
-          disable = if (cfg.disabledLanguages != [ ]) then cfg.disabledLanguages else null;
+  # NOTE: We call setup manually below.
+  callSetup = false;
+  # NOTE: We install cfg.package manually so we can install grammars using it.
+  installPackage = false;
 
-          custom_captures = if (cfg.customCaptures != { }) then cfg.customCaptures else null;
-        };
+  extraConfig = cfg: {
+    warnings =
+      concatMap
+        (
+          name:
+          optional (!cfg.nixGrammars && cfg.${name} == null) ''
+            Nixvim (plugins.treesitter): You have disabled `nixGrammars`, but `plugins.treesitter.${name}` is `null`.
+            This package is required to build the grammars at runtime.
+          ''
+        )
+        [
+          "gccPackage"
+          "nodejsPackage"
+          "treesitterPackage"
+        ];
```

Comment:
I don't think we should have those assertions, there are very good reasons to have them null with nixGrammars = false, I'd say it's even the reason they are nullable.

The main reason is that you may want to use the C compiler on your PATH, to find required libraries for example. We had an issue that required this change in the first place, as using this gcc will make other stuff break, for example the `cc` in the neovim terminal may use a bad compiler.

Reply:
(Because setting them to null is an explicit action from the user if I understand the code correctly)

Reply:
> I don't think we should have those assertions

As these are warnings I don't think it's so bad? We did have assertions before but they were switched to warnings for the kind've reasons you describe.

Is your issue that the user cannot silence the warnings?

Reply:
Yeah exactly

Reply:
Although I think I was the one to suggest assertions/warnings in the first place, after discussing on matrix we've concluded they aren't worth it in this case:
- They will only show for users who have explicitly disabled these packages
- Such users should know what they're doing
- If we had warnings, there'd be no practical way for users to silence them
Therefore the cons outweigh the pros.

Reply:
I agree with this conclusion (in case someone cares).